File: client/go/connection.go
package rendis

import (
	"net/http"

	"github.com/gorilla/websocket"
)

type Client struct {
	conn *websocket.Conn
}

func New(url, key string) (*Client, error) {
	header := http.Header{}
	header.Set("X-RENDIS-Key", key)

	conn, _, err := websocket.DefaultDialer.Dial(url+"/connect", header)
	if err != nil {
		return nil, err
	}

	return &Client{
		conn: conn,
	}, nil
}

func (c *Client) Close() error {
	return c.conn.Close()
}

func (c *Client) send(data []byte) error {
	return c.conn.WriteMessage(websocket.BinaryMessage, data)
}

func (c *Client) receive() ([]byte, error) {
	_, data, err := c.conn.ReadMessage()
	if err != nil {
		return nil, err
	}

	return data, nil
}

File: client/python/rendis/connection.py
import websocket

class ClientConnection:
    def __init__(self, url: str, key: str):
        connect_url = url + "/connect"
        # Using websocket-client to connect synchronously
        self.conn = websocket.create_connection(
            connect_url,
            header=[f"X-RENDIS-Key: {key}"]
        )

    def close(self):
        self.conn.close()

    def send(self, data: bytes):
        self.conn.send_binary(data)

    def receive(self) -> bytes:
        return self.conn.recv()
